src/refresh/RefreshTimer.ts

```diff
@@ -0,0 +1,103 @@
+ *  - Because of the jitter, the maximum backoff time is actually `MaximumBackoffInMs` * (1 + `JitterRatio`).
+ */
+
+const MinimumBackoffInMs = 30 * 1000; // 30s
+const MaximumBackoffInMs = 10 * 60 * 1000; // 10min
+const MaxSafeExponential = 53; // Used to avoid overflow, as Number.MAX_SAFE_INTEGER = 2^53 - 1.
+const JitterRatio = 0.25;
+
+export class RefreshTimer {
+    private _minBackoff: number = MinimumBackoffInMs;
+    private _maxBackoff: number = MaximumBackoffInMs;
+    private _failedAttempts: number = 0;
+    private _backoffEnd: number; // Timestamp
+    constructor(
+        private _interval: number
+    ) {
+        if (this._interval <= 0) {
+            throw new Error(`Refresh interval must be greater than 0. Given: ${this._interval}`);
+        }
+
+        this._backoffEnd = Date.now() + this._interval;
+    }
+
+    public canRefresh(): boolean {
+        return Date.now() >= this._backoffEnd;
+    }
+
+    public backoff(): void {
+        this._backoffEnd = Date.now() + this._calculateBackoffTime();
+        this._failedAttempts += 1;
+    }
+
+    public reset(): void {
+        this._backoffEnd = Date.now() + this._interval;
+        this._failedAttempts = 0;
+    }
+
+    private _calculateBackoffTime(): number {
+        let minBackoffMs: number;
+        let maxBackoffMs: number;
+        if (this._interval <= this._minBackoff) {
+            return this._interval;
+        }
+
+        // _minBackoff <= _interval
+        if (this._interval <= this._maxBackoff) {
+            minBackoffMs = this._minBackoff;
+            maxBackoffMs = this._interval;
+        } else {
+            minBackoffMs = this._minBackoff;
+            maxBackoffMs = this._maxBackoff;
+        }
+
+        // exponential: minBackoffMs * 2^failedAttempts
+        const exponential = Math.min(this._failedAttempts, MaxSafeExponential);
+        let calculatedBackoffMs = minBackoffMs * (efficientPowerOfTwo(exponential));
+        if (calculatedBackoffMs > maxBackoffMs) {
+            calculatedBackoffMs = maxBackoffMs;
+        }
+
+        // jitter: random value between [-1, 1) * jitterRatio * calculatedBackoffMs
+        const jitter = JitterRatio * (Math.random() * 2 - 1);
+
+        return calculatedBackoffMs * (1 + jitter);
+    }
+
+}
+
+/**
+ * Efficient way to calculate 2^exponential.
+ *
+ * `Math.pow(base, exp)` is not used because it is less efficient and accurate by operating floating-point number.
+ * `1 << exp` is not used because it returns wrong results when exp >= 31.
+ */
+function efficientPowerOfTwo(positiveExponential: number) {
+    if (positiveExponential < 0) {
+        throw new Error("exponential must be a non-negative integer.");
+    } else if (positiveExponential > MaxSafeExponential) {
+        throw new Error(`exponential must be less than or equal to ${MaxSafeExponential}.`);
+    }
+
+    // bitwise operations in JavaScript are limited to 32 bits. It overflows at 2^31 - 1.

```

Comment:
What do you think about changing the `MaxSafeExponential` to 30 then? In reality, we wouldn't reach that high number of attempts anyway.

Reply:
The max backoff will always be less than 10 minutes. That means, `minBackoffMs * efficientPowerOfTwo` will always be capped at 600,000 ms (10 minutes). `2^20` already goes much higher than 600k, so we know that we would only backoff for 10 minutes after 19 failed attempts (max).

Reply:
The number was determined by the language’s limitation, for safe operations. I don’t want it to be determined by the max backoff time, otherwise they are coupled too much and it’s risky to change max backoff time(just in case).

But yes, 30 is reasonable, as it’s determined by the bitwise operation limitation. It’s large enough for reasonable backoff as you mentioned. And the benefit is we don’t need the efficientPowerOfTwo utility any more.

Reply:
updated in 98b12e2